pkg/addonmanager/controllers/managementaddonconfig/controller.go

```diff
@@ -0,0 +1,314 @@
+	clusterManagementAddonInformers addoninformerv1alpha1.ClusterManagementAddOnInformer,
+	configInformerFactory dynamicinformer.DynamicSharedInformerFactory,
+	configGVRs map[schema.GroupVersionResource]bool,
+) factory.Controller {
+	syncCtx := factory.NewSyncContext(controllerName)
+
+	c := &clusterManagementAddonConfigController{
+		addonClient:                   addonClient,
+		clusterManagementAddonLister:  clusterManagementAddonInformers.Lister(),
+		clusterManagementAddonIndexer: clusterManagementAddonInformers.Informer().GetIndexer(),
+		configListers:                 map[string]dynamiclister.Lister{},
+		queue:                         syncCtx.Queue(),
+	}
+
+	configInformers := c.buildConfigInformers(configInformerFactory, configGVRs)
+
+	if err := clusterManagementAddonInformers.Informer().AddIndexers(cache.Indexers{byClusterManagementAddOnConfig: c.indexByConfig}); err != nil {
+		utilruntime.HandleError(err)
+	}
+
+	return factory.New().
+		WithSyncContext(syncCtx).
+		WithInformersQueueKeysFunc(func(obj runtime.Object) []string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return []string{key}
+		}, clusterManagementAddonInformers.Informer()).
+		WithBareInformers(configInformers...).
+		WithSync(c.sync).ToController(controllerName)
+}
+
+func (c *clusterManagementAddonConfigController) buildConfigInformers(
+	configInformerFactory dynamicinformer.DynamicSharedInformerFactory,
+	configGVRs map[schema.GroupVersionResource]bool,
+) []factory.Informer {
+	configInformers := []factory.Informer{}
+	for gvr := range configGVRs {
+		indexInformer := configInformerFactory.ForResource(gvr).Informer()
+		_, err := indexInformer.AddEventHandler(
+			cache.ResourceEventHandlerFuncs{
+				AddFunc: c.enqueueClusterManagementAddOnsByConfig(gvr),
+				UpdateFunc: func(oldObj, newObj interface{}) {
+					c.enqueueClusterManagementAddOnsByConfig(gvr)(newObj)
+				},
+				DeleteFunc: c.enqueueClusterManagementAddOnsByConfig(gvr),
+			},
+		)
+		if err != nil {
+			utilruntime.HandleError(err)
+		}
+		configInformers = append(configInformers, indexInformer)
+		c.configListers[toListerKey(gvr.Group, gvr.Resource)] = dynamiclister.New(indexInformer.GetIndexer(), gvr)
+	}
+	return configInformers
+}
+
+func (c *clusterManagementAddonConfigController) enqueueClusterManagementAddOnsByConfig(gvr schema.GroupVersionResource) enqueueFunc {
+	return func(obj interface{}) {
+		name, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("error to get accessor of object: %v", obj))
+			return
+		}
+
+		objs, err := c.clusterManagementAddonIndexer.ByIndex(byClusterManagementAddOnConfig, fmt.Sprintf("%s/%s/%s", gvr.Group, gvr.Resource, name))
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("error to get addons: %v", err))
+			return
+		}
+
+		for _, obj := range objs {
+			if obj == nil {
+				continue
+			}
+
+			cma := obj.(*addonapiv1alpha1.ClusterManagementAddOn)
```

Comment:
you do not need this line for MetaNamespaceKeyFunc, or you want to log an error when type is not cma?